### PR TITLE
Support for Flow's Optional Indexed Access type

### PR DESCRIPTION
**Pull request by gkz**
## Description

This adds support for Flow's "Optional Indexed Access Types" (e.g `type T = Obj?.[K]`), a follow on to the support for "Indexed Access Types" added in https://github.com/prettier/prettier/pull/10594

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by gkz** [comment]
Failure doesn't look related.

File: tests/format/flow/optional-indexed-access/jsfmt.spec.js
Comment (fisker) [inline tests/format/flow/optional-indexed-access/jsfmt.spec.js:1]: `babel-flow` doesn't support it, correct?

Comment (sosukesuzuki) [reply tests/format/flow/optional-indexed-access/jsfmt.spec.js:1]: Maybe Babel 7.14.0 supports it, but we haven't updated yet

File: src/language-js/print/flow.js
Comment (fisker) [inline src/language-js/print/flow.js:90]: Let's move the `optional` check to function, so we can keep this file small.

File: tests/format/flow/optional-indexed-access/jsfmt.spec.js
Comment (fisker) [reply tests/format/flow/optional-indexed-access/jsfmt.spec.js:1]: If that's true, better update babel first.

Comment (gkz) [reply tests/format/flow/optional-indexed-access/jsfmt.spec.js:1]: Babel 7.14.0 does support it (https://github.com/babel/babel/releases/tag/v7.14.0), didn't realize it had released (happened just today)

File: src/language-js/print/flow.js
Comment (gkz) [reply src/language-js/print/flow.js:90]: Only thing is that `printIndexedAccessType` is shared between Flow and TypeScript, and while both of them have indexed access types, but only Flow supports optional indexed access types. Is that OK?

Comment (fisker) [reply src/language-js/print/flow.js:90]: Yes, we can use `node.type === 'OptionalIndexedAccessType' && node.optional`, make it strict.

File: tests/format/flow/optional-indexed-access/jsfmt.spec.js
Comment (sosukesuzuki) [reply tests/format/flow/optional-indexed-access/jsfmt.spec.js:1]: However probably just update to 7.14,  tests will be failed. Because almost error messages thrown by babel/parser are updated.

Comment (sosukesuzuki) [reply tests/format/flow/optional-indexed-access/jsfmt.spec.js:1]: I'll create PR to update babel/parser

Comment (fisker) [reply tests/format/flow/optional-indexed-access/jsfmt.spec.js:1]: I'm not doing anything now, I can try.

Comment (sosukesuzuki) [reply tests/format/flow/optional-indexed-access/jsfmt.spec.js:1]: Sorry I've created https://github.com/prettier/prettier/pull/10802

Comment (fisker) [reply tests/format/flow/optional-indexed-access/jsfmt.spec.js:1]: I was busy... suddenly

**Comment by gkz** [comment]
If there are issues with the Babel upgrade, we can merge this PR and just add `babel-flow` to this test (and also the `tests/format/flow/indexed-access` test) in the Babel PR

**Review by sosukesuzuki — APPROVED** [review]
We can support indexed access type syntax by `babel-flow` after #10802 is merged. So this PR looks good to me.

**Comment by sosukesuzuki** [comment]
I'll merge this for now, if you have another review points please let me know.

**Pull request merged**
